rcl/test/rcl/test_events.cpp

```diff
@@ -370,56 +435,50 @@ TEST_F(CLASSNAME(TestEventFixture, RMW_IMPLEMENTATION), test_pubsub_liveliness_k
+  conditional_wait_for_msgs_and_events<rmw_liveliness_changed_status_t, void>(
+    conditional_func, &subscription, &subscription_event,
+    nullptr, context_ptr, 1000,
+    &msg_persist_ready, &subscription_persist_ready, &publisher_persist_ready,
+    &msg, &liveliness_status, nullptr);
 
-  bool msg_ready, subscription_event_ready, publisher_event_ready;
-  bool msg_persist_ready = false;
-  bool subscription_persist_ready = false;
-  bool publisher_persist_ready = false;
-  auto timeout = milliseconds(2000);
-  auto start_time = std::chrono::system_clock::now();
-  do {
-    // wait for msg and events
-    wait_for_msgs_and_events(&subscription, &subscription_event, nullptr,
-      context_ptr, 1000, &msg_ready, &subscription_event_ready, &publisher_event_ready);
-    if (msg_ready) {
-      test_msgs__msg__Primitives msg;
-      test_msgs__msg__Primitives__init(&msg);
-      OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT({
-        test_msgs__msg__Primitives__fini(&msg);
-      });
-      ret = rcl_take(&subscription, &msg, nullptr);
-      EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
-      EXPECT_EQ(std::string(msg.string_value.data, msg.string_value.size),
-        std::string(test_string));
-    }
-    if (subscription_event_ready) {
-      rmw_liveliness_changed_status_t liveliness_status;
-      ret = rcl_take_event(&subscription_event, &liveliness_status);
-      EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
-      EXPECT_EQ(liveliness_status.alive_count, 0);
-      // TODO(ross-desmond): Connext and OpenSplice seem to be counting liveliness differently
-      if (is_opensplice) {
-        EXPECT_EQ(liveliness_status.alive_count_change, 2);
-      } else {
-        EXPECT_EQ(liveliness_status.alive_count_change, 0);
-      }
-      EXPECT_EQ(liveliness_status.not_alive_count, 0);
-      EXPECT_EQ(liveliness_status.not_alive_count_change, 0);
-    }
-    msg_persist_ready |= msg_ready;
-    subscription_persist_ready |= subscription_event_ready;
-    publisher_persist_ready |= publisher_event_ready;
-    if (std::chrono::system_clock::now() - start_time > timeout) {
-      break;
-    }
-  } while (!(msg_persist_ready && subscription_persist_ready));
   // test that the message published to topic is as expected
   EXPECT_TRUE(msg_persist_ready);
   // test subscriber/datareader liveliness changed status
   EXPECT_TRUE(subscription_persist_ready);
   // test that the killed publisher/datawriter has no active events
   EXPECT_FALSE(publisher_persist_ready);
 
+  if (msg_persist_ready) {
+    EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+    EXPECT_EQ(std::string(msg.string_value.data, msg.string_value.size),
+      std::string(test_string));
+  }
+  if (subscription_persist_ready) {
+    EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+    EXPECT_EQ(liveliness_status.alive_count, 0);
+    // TODO(ross-desmond): Connext and OpenSplice seem to be counting liveliness differently
```

Comment:
This is concerning... I'd actual say that opensplice is just broken (or maybe Connext is?). Do you have a reason that they might be disagreeing?

Reply:
@ross-desmond, @nburek, and I believe this is a case of "distance vs. displacement". In this test, the publisher comes alive (`alive_count` +1) and is then shut down (`alive_count` -1). Although `alive_count` ends up where it started (at 0), that's two state changes to `alive_count`. So it looks like Connext's `alive_count_change` represents the literal change to `alive_count` while OpenSplice's `alive_count_change` represents the number of state changes to `alive_count`.

Reply:
The spec reads that `alive_count_change` is "The change in the alive_count since the last time the listener was called or the status was read". Our interpretation of that statement leads us to believe that Connext's behavior is the correct one.

Reply:
Sounds like a potential bug in OpenSplice then, or at least it would be nice to know if this is intentional or not. Can one of you summarize this issue on OpenSplice's issue tracker and ask for their opinion on this?

https://github.com/ADLINK-IST/opensplice/issues

/cc @dirk-thomas

Reply:
For now we can continue with the exception in the test and a TODO, but ideally we'd come back and address this with a definitive comment that they disagree intentionally, or that we adjust for the difference of opinion in our layer, or something like that.

Reply:
I have opened an issue on OpenSplice's issue tracker here: https://github.com/ADLINK-IST/opensplice/issues/88.

Thanks.

Reply:
Thank you!